### PR TITLE
test/fuzz: add fuzzing tests for xrow_header_decode

### PR DESCRIPTION
**Pull request by ligurio**
NO_DOC=testing
NO_CHANGELOG=testing

### CODE REVIEW

File: test/fuzz/xrow_header_decode_fuzzer.c [hunk 1]
```diff
@@ -0,0 +1,38 @@
+#include "box/iproto_constants.h"
+#include "box/xrow.h"
+#include "trivia/util.h"
+#include "fiber.h"
+#include "memory.h"
+
+void
+cord_on_yield(void) {}
```
Comment (sergos) [inline test/fuzz/xrow_header_decode_fuzzer.c:8]: Are this one and the fiber init are caused by the luajit build triggered by xrow library?
Can it be evaded in xrow dependency?

**Review by sergos — COMMENTED** [review]
Just some questions, LGTM otherwise.

**Review by sergos — APPROVED** [review]

File: test/fuzz/xrow_header_decode_fuzzer.c (hunk 1 above)
Comment (ligurio) [reply test/fuzz/xrow_header_decode_fuzzer.c:6]: > Are this one and the fiber init are caused by the luajit build triggered by xrow library?

As I got it right, it is triggered by libcore.

> Can it be evaded in xrow dependency?

Seems no:

```
$ make VERBOSE=1 xrow_header_decode_fuzzer

<snipped>

/bin/clang++  -fexceptions -funwind-tables -fasynchronous-unwind-tables -fno-common -msse2 -fsanitize=fuzzer-no-link -fsanitize=address -fsanitize-blacklist=/home/sergeyb/sources/MRG/tarantool/asan/asan.supp  -fmacro-prefix-map=/home/sergeyb/sources/MRG/tarantool=. -std=c++11 -Wall -Wextra -Wno-invalid-offsetof -Wno-gnu-alignof-expression -Wno-cast-function-type -Werror -Wno-unused-parameter -g -ggdb -O0 -rdynamic CMakeFiles/xrow_header_decode_fuzzer.dir/xrow_header_decode_fuzzer.c.o -o xrow_header_decode_fuzzer  ../../src/box/libxrow.a ../../src/libserver.a ../../src/lib/coll/libcoll.a ../../src/lib/http_parser/libhttp_parser.a ../../src/lib/swim/libswim.a ../../src/lib/swim/libswim_udp.a ../../src/lib/swim/libswim_ev.a ../../src/lib/crypto/libcrypto.a ../../src/libcrc32.a ../../src/libcpu_feature.a ../../src/box/libbox_error.a ../../src/lib/vclock/libvclock.a ../../src/lib/mpstream/libmpstream.a ../../src/libstat.a ../../src/lib/core/libcore.a ../../src/lib/small/libsmall.a ../../libmisc.a ../../src/lib/bit/libbit.a ../../src/lib/uri/liburi.a ../../src/lib/tzcode/libtzcode.a ../../src/lib/salad/libsalad.a ../../libdecNumber.a ../../libev.a -lm ../../libeio.a -lpthread ../../libcoro.a -licui18n -licuuc -licudata ../../third_party/c-dt/build/libcdt.a -lssl -lcrypto -ldl ../../third_party/libunwind/src/.libs/libunwind-x86_64.a ../../third_party/libunwind/src/.libs/libunwind.a ../../src/box/libnode_name.a ../../src/lib/msgpuck/libmsgpuck.a -fsanitize=fuzzer -fsanitize-coverage=trace-cmp -fprofile-instr-generate -fcoverage-mapping -g 
/bin/ld: /bin/ld: DWARF error: invalid or unhandled FORM value: 0x23
../../src/lib/core/libcore.a(fiber.c.o): in function `fiber_call':
fiber.c:(.text.fiber_call[fiber_call]+0x3a1): undefined reference to `cord_on_yield'
/bin/ld: ../../src/lib/core/libcore.a(fiber.c.o): in function `fiber_yield_impl':
fiber.c:(.text.fiber_yield_impl[fiber_yield_impl]+0x3cf): undefined reference to `cord_on_yield'
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
make[3]: *** [test/fuzz/CMakeFiles/xrow_header_decode_fuzzer.dir/build.make:132: test/fuzz/xrow_header_decode_fuzzer] Error 1
make[3]: Leaving directory '/home/sergeyb/sources/MRG/tarantool/build'
make[2]: *** [CMakeFiles/Makefile2:8660: test/fuzz/CMakeFiles/xrow_header_decode_fuzzer.dir/all] Error 2
make[2]: Leaving directory '/home/sergeyb/sources/MRG/tarantool/build'
make[1]: *** [CMakeFiles/Makefile2:8667: test/fuzz/CMakeFiles/xrow_header_decode_fuzzer.dir/rule] Error 2
make[1]: Leaving directory '/home/sergeyb/sources/MRG/tarantool/build'
make: *** [Makefile:3193: xrow_header_decode_fuzzer] Error 2
```

`memory_init` and `memory_free` are required for slab cache initialization. Without calling these functions, assertion is triggered on starting test:

```
INFO: A corpus is not provided, starting from an empty corpus                                                                                
xrow_header_decode_fuzzer: src/lib/small/small/slab_cache.c:178: void slab_cache_create(struct slab_cache *, struct slab_arena *): Assertion `(long)arena->slab_size >= min_order0_size' failed.
==75511== ERROR: libFuzzer: deadly signal 
```

but `fiber_init` and `fiber_free` could be removed:

```diff
--- a/test/fuzz/xrow_header_decode_fuzzer.c
+++ b/test/fuzz/xrow_header_decode_fuzzer.c
@@ -1,7 +1,5 @@
 #include "box/iproto_constants.h"
 #include "box/xrow.h"
-#include "trivia/util.h"
-#include "fiber.h"
 #include "memory.h"
 
 void
@@ -12,14 +10,12 @@ static void
 setup(void)
 {
 	memory_init();
-	fiber_init(fiber_c_invoke);
 }
 
 __attribute__((destructor))
 static void
 teardown(void)
 {
-	fiber_free();
 	memory_free();
 }
```

Comment (sergos) [reply test/fuzz/xrow_header_decode_fuzzer.c:6]: Very good.
And what about the `cord_on_yield` since it is also relevant to the fiber machinery?

Comment (ligurio) [reply test/fuzz/xrow_header_decode_fuzzer.c:6]: It is required for building, without it, build failed (see log in previous comment).

**Review by ochaplashkin — APPROVED** [review]

**Comment by igormunkin** [comment]
Also backported to 2.11 and 2.10 in scope of 173f8ed and 5e2a0b4 respectively.

**Pull request merged**